### PR TITLE
Small issues

### PR DESCRIPTION
**Pull request by jgozal**
This PR addresses #58 and #41, but 95% of the code is about #57. Also made some minor improvements to the summary here https://github.com/jgozal/moneycompass/pull/61/commits/4c873af6c0453eb5d5268250f1c85d482ee05adb. I am not very satisfied with my implementation for #57 . There's a bit of code duplication and things aren't as organized and straightforward as I would like them to be, but it works. As always, any feedback is welcome. 

Please pull the branch and play with the switches. I've added 2 in the input card, and 2 more on top of the amortization table. I think they improve UX a little.  Also read all comments below in this page. Most are outdated as I've made some recent changes, but still relevant. Sorry for the heavy PR, hiding white space when reviewing will help.

image

### CODE REVIEW

File: src/utils/numberFormat.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import numbro from 'numbro'
+import _ from 'lodash'
 
-export const toUSD = value => {
-  return numbro(value).formatCurrency({
-    thousandSeparated: true,
-    mantissa: 0
+export const toUSD = (value, mantissa) => {
+  return numbro(_.round(value, mantissa)).formatCurrency({
```
Comment (jgozal) [inline src/utils/numberFormat.js]: Related to #41 . Now the 30th year in the amortization table finishes with 0 as opposed to -5, and numbers in the summary make sense as well.

File: src/components/mortgage-investment-compare/amortization-table.js [hunk 2]
```diff
@@ -12,7 +12,7 @@ const tableDivider = css`
 `
 const fixedTableLayout = css`
   display: table;
-  width: 100%;
+  width: 53.5rem;
```
Comment (jgozal) [inline src/components/mortgage-investment-compare/amortization-table.js]: I had to give the table a fixed width. Let me know how it looks on your end. This should solve the issue with the thead and the tbody being misaligned but it has a secondary minor side effect which I prefer: the scrollbar will be separated from the tbody when you zoom out. We'll come up with a better implementation of this table some time in the future but for now I think this works well

File: src/components/mortgage-investment-compare/amortization-table.js [hunk 3]
```diff
@@ -184,6 +204,14 @@ const AmortizationTable = props => {
           })}
         </Tbody>
       </Table>
+      {props.includeInflation && (
+        <small>
+          <b>
+            ** If you include inflation, your payments will never actually
+            become smaller; they will have a smaller purchasing power.
+          </b>
+        </small>
```
Comment (jgozal) [inline src/components/mortgage-investment-compare/amortization-table.js]: small disclaimer I added at the bottom of the table. Chris told me that it would be confusing to some people to see their payments decrease, because that's not what's actually happening, so just clarifying. Only appears when you include inflation.

File: src/components/mortgage-investment-compare/index.js [hunk 4]
```diff
@@ -49,17 +61,41 @@ class MortgageInvestmentCompare extends React.Component {
 
     const result = this.getResult(input)
```
Comment (jgozal) [inline src/components/mortgage-investment-compare/index.js]: I had to move this to `componentWillMount` because I was having some sort of race condition that prevented me from  reading `includeInflation` and `includeROI` from the state when launching the app. Feel free to pull locally and revert the change if you'd like to dig further into why this was happening.

File: src/components/mortgage-investment-compare/index.js [hunk 5]
```diff
@@ -49,17 +61,41 @@ class MortgageInvestmentCompare extends React.Component {
...
+    this.setState({ includeROI: !this.state.includeROI }, () => {
+      const input = _.cloneDeep(this.state.input)
+      input.investmentRate = this.state.includeROI
+        ? this.state.input.investmentRate
+        : 0
+
+      const result = this.getResult(input)
+      this.setState({ result })
+    })
+  }
```
Comment (jgozal) [inline src/components/mortgage-investment-compare/index.js]: fyi, `setState` takes a callback as a second argument which is called after setting the state. So I'm setting the state and changing the boolean for `includeROI`, and then after that is done, running `getResult` and setting state again. I don't particularly love this implementation but it works.

File: src/components/mortgage-investment-compare/index.js [hunk 6]
```diff
@@ -76,33 +112,30 @@ class MortgageInvestmentCompare extends React.Component {
 
   // Returns future value dynamically depending on mortgage term length.
   calculateFV (option1, option2, input) {
+    const investmentRate = this.state.includeROI
+      ? input.investmentRate / 100 / COMPOUND_FREQUENCY
+      : 0
+    const inflationRate = this.state.includeInflation
+      ? input.inflation / 100
+      : 0
+
```
Comment (jgozal) [inline src/components/mortgage-investment-compare/index.js]: I pretty much do this same thing a little below in `getResult`. Not a fan of code duplication and tbh, we don't really need `calculateFV` anymore as we can just pull the final FV values from the amortization table, but I guess it's nice to have for now to double check our work. I'm sure we'll eventually get rid of the bulk calc.

File: src/components/mortgage-investment-compare/input-card.js [hunk 7]
```diff
@@ -197,13 +198,24 @@ const InputCard = props => {
...
+              id='investmentRate'
+            >
+              <label>Return on Investment (ROI)</label>
+            </Tooltip>
+            <CustomInput
+              type='checkbox'
+              className='custom-switch'
+              id='roi-switch'
+              name='roi-switch'
+              label={<small>Include?</small>}
```
Comment (jgozal) [inline src/components/mortgage-investment-compare/input-card.js]: I used `<small>` but still not 100% convinced with the font styling. If you can make it better please be my guest. If you like it like this then nbd.

File: src/components/mortgage-investment-compare/amortization-table.js [hunk 8]
```diff
@@ -133,8 +133,28 @@ const hoverTableCells = (year, option1, option2) => {
...
+        />
+        <CustomInput
+          type='checkbox'
+          className='custom-switch ml-4'
+          id='inflation-switch-yearly-breakdown'
+          name='inflation-switch-yearly-breakdown'
+          label={<small>Include Inflation?</small>}
+          checked={props.includeInflation}
+          onChange={props.onInflationSwitch}
+        />
```
Comment (jgozal) [inline src/components/mortgage-investment-compare/amortization-table.js]: should the switches be light blue like our links and feedback button?

File: src/components/mortgage-investment-compare/index.js (hunk 5 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/index.js]: I'm going to change this a little. I'll be using lifecycle methods instead of callbacks. Will push my changes within the next hour

Comment (jgozal) [reply src/components/mortgage-investment-compare/index.js]: took care of it here: https://github.com/jgozal/moneycompass/pull/61/commits/052412c46b114690bf0fe7a5c315d77adaf65a81

File: package.json [hunk 9]
```diff
@@ -47,7 +47,6 @@
     "bootstrap": "^4.3.1",
     "emotion": "^9.1.3",
     "formulajs": "^1.0.8",
-    "jquery": "^1.9.1",
     "lodash": "^4.17.15",
```
Comment (jgozal) [inline package.json]: gh marks it as a security vulnerability and we don't really need it

File: src/components/mortgage-investment-compare/amortization-table.js [hunk 10]
```diff
@@ -12,7 +12,7 @@ const tableDivider = css`
 `
 const fixedTableLayout = css`
   display: table;
-  width: 100%;
+  width: ${window.innerWidth * 0.595}px;
```
Comment (jgozal) [inline src/components/mortgage-investment-compare/amortization-table.js]: Related to #58. I had to give the table a fixed width. Let me know how it looks on your end. This should solve the issue with the thead and the tbody being misaligned but it has a secondary minor side effect which I prefer: the scrollbar will be separated from the tbody when you zoom out. We'll come up with a better implementation of this table some time in the future but for now I think this works well

File: package.json (hunk 9 above)
Comment (hdavidzhu) [reply package.json]: haha what did we use it for originally?

File: src/components/mortgage-investment-compare/amortization-table.js (hunk 10 above)
Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/amortization-table.js]: I think it looks good:

image

File: src/components/mortgage-investment-compare/amortization-table.js [hunk 11]
```diff
@@ -128,64 +128,113 @@ const hoverTableCells = (year, option1, option2) => {
...
-  )
+class AmortizationTable extends React.Component {
+  componentWillReceiveProps (nextProps) {
+    if (
+      this.props.includeROI !== nextProps.includeROI ||
+      this.props.includeInflation !== nextProps.includeInflation
+    ) {
+      this.props.updtateResult(nextProps.input)
+    }
+  }
```
Comment (hdavidzhu) [inline src/components/mortgage-investment-compare/amortization-table.js]: Was this a performance improvement?

File: src/components/mortgage-investment-compare/index.js [hunk 12]
```diff
@@ -23,16 +23,28 @@ import {
...
-  constructor () {
-    super()
+  constructor (props) {
+    super(props)
     this.calculateFV = this.calculateFV.bind(this)
     this.calculateInterestAmt = this.calculateInterestAmt.bind(this)
     this.calculateOpportunityCost = this.calculateOpportunityCost.bind(this)
     this.calculatePMT = this.calculatePMT.bind(this)
-    this.getResult = this.getResult.bind(this)
+    this.updtateResult = this.updtateResult.bind(this)
```
Comment (hdavidzhu) [inline src/components/mortgage-investment-compare/index.js]:
```suggestion
    this.updateResult = this.updateResult.bind(this)
```

Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/index.js]: One issue w/ this is that `updateResult` now becomes stateful. With `getResult`, we never had to worry about side effects w/ the app since it's pure. When we needed to modify state, we did so explicitly in a single place.

File: src/components/mortgage-investment-compare/index.js (hunk 5 above)
Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/index.js]: What was happening before you made the optimization?

File: src/components/mortgage-investment-compare/index.js [hunk 13]
```diff
@@ -174,12 +194,17 @@ class MortgageInvestmentCompare extends React.Component {
...
+            each situation, with the same budget.
           </p>
           <InputCard
             onInputChange={this.updateInput}
+            onROISwitch={this.toggleIncludeROI}
+            onInflationSwitch={this.toggleIncludeInflation}
+            includeInflation={this.state.includeInflation}
+            includeROI={this.state.includeROI}
             input={this.state.input}
+            updtateResult={this.updtateResult}
```
Comment (hdavidzhu) [inline src/components/mortgage-investment-compare/index.js]: Why does the input card need to update the result directly?

File: src/components/mortgage-investment-compare/input-card.js [hunk 14]
```diff
@@ -18,328 +19,374 @@ import Accordion from '../general/accordion'
...
+
+  componentWillReceiveProps (nextProps) {
+    if (
+      this.props.input !== nextProps.input ||
+      this.props.includeROI !== nextProps.includeROI ||
+      this.props.includeInflation !== nextProps.includeInflation
+    ) {
+      this.props.updtateResult(nextProps.input)
+    }
+  }
```
Comment (hdavidzhu) [inline src/components/mortgage-investment-compare/input-card.js]: Why is this needed?

File: src/components/mortgage-investment-compare/index.js (hunk 5 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/index.js]: @hdavidzhu I was calling `getResult` and setting the state for a second time as a callback to `setState`. Not really the best way to do it

File: package.json (hunk 9 above)
Comment (jgozal) [reply package.json]: no idea, probs something we did in the very beginning to manipulate the DOM

File: src/components/mortgage-investment-compare/amortization-table.js (hunk 10 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/amortization-table.js]: great. Just curious, why isnt' the scrollbar there? is your screen that big? 

image

File: src/components/mortgage-investment-compare/amortization-table.js (hunk 11 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/amortization-table.js]: I didn't run a profiler, but it's likely insignificant

File: src/components/mortgage-investment-compare/index.js (hunk 13 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/index.js]: just like it updates the input, please see long answer below

File: src/components/mortgage-investment-compare/input-card.js (hunk 14 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/input-card.js]: @hdavidzhu it seems like there are 4-5 questions/comments about this so I will try to address all of them here and explain the thought process.

What's happening now is that we are adding 2 new properties to the state: `includeROI` and `includeInflation`. We need to listen for changes in these and update the results every time the user enables/disables inflation. This has always been done with lifecycle methods in React, and now more recently with hooks.

1. Hooks: hooks are great but changing `MortgageInvestmentCompare` to a function would have required significant work so I decided to not do it this way.

2. Lifecycle methods: historically, `componentDidUpdate` has always been the default lifecycle method to monitor changes in the state within a component. I could have used `componentDidUpdate` in `MortgageInvestmentCompare` to monitor changes in `includeROI`  and `includeInflation`, but `componentDidUpdate` is now deprecated. So I decided to use `componentWillReceiveProps` instead which monitors changes in the props being passed to the child components. 

I also initially decided to call `getResults` within `setState`'s callback as explained above, but that's a very "dirty and quick" way of doing it.

Because we are now monitoring for changes in the parent's state within the child components, we need to pass in a function that changes the state in the parent whenever changes are detected. As you said, `getResults` was purse, so calling this in the child component and setting state there wouldn't have worked as it would set the state on the child component as well. While your concerns about the function becoming stateful now make sense, before we were setting the state in the `updateInput` handler like so:

```
updateInput (input) {
    const result = this.getResult(input)
    this.setState({ input, result })
  }
```

Which isn't really a great approach since the goal of `updateInput` is to update the input, and I don't find it particularly dangerous to have an `updateResult`.  What we do with `updateResult` right now is no different from what we do with `updateInput` as `updateInput` also gets called within the child component through the onChange handlers, and runs in the parent to change the state. Passing more responsibility and information to the child components is completely fine and normal.  If you'd prefer it, we could go back to using `getResult` to keep it pure, and then we can create a new method `updateResult` which only sets the state with whatever `getResult` returns. 

The current approach is probably not the cleanest and perhaps not as straightforward for someone that hasn't used React for some time, but it's safe and gets the job done. In the future, we should consider using the new hooks which would have worked great in this situation. Converting everything we have right now to use hooks instead wouldn't be that hard but is unnecessary at this point.

File: src/components/mortgage-investment-compare/index.js (hunk 12 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/index.js]: please see long comment below

File: src/components/mortgage-investment-compare/amortization-table.js (hunk 10 above)
Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/amortization-table.js]: Scrollbars don't show up on MacOS I don't think

File: src/components/mortgage-investment-compare/input-card.js (hunk 14 above)
Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/input-card.js]: > What's different now is that results are not affected only by changes in the input's on change handler so this also doesn't get the job done.

What if you handled that in the parent component by updating the results when you detect changes in the input?

> Converting everything we have right now to use hooks wouldn't be that hard but is unnecessary at this point.

I think it's pretty straightforward to get away w/o using it. Since pairing is tricky for us, I'll spin up a temporary PR to demonstrate what I mean

Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/input-card.js]: @jgozal Here's what I had in mind: https://github.com/jgozal/moneycompass/pull/62

Give it a run and lemme know what you think

File: src/components/mortgage-investment-compare/amortization-table.js (hunk 10 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/amortization-table.js]: so do you just scroll without the scrollbar basically?

Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/amortization-table.js]: Yep! Using the touchpad. The scroll comes up when I scroll:

image

File: src/components/mortgage-investment-compare/input-card.js (hunk 14 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/input-card.js]: @hdavidzhu here it is https://github.com/jgozal/moneycompass/pull/61/commits/239f358a8fad676ed6ccf7e013fa2cb1625f05fc

Making another small commit to fix old docs

File: src/components/mortgage-investment-compare/amortization-table.js (hunk 8 above)
Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/amortization-table.js]: if possible but nbd

I think in Sass we would have defined our standard colors w/ some variables and that should make all colors consistent to our theme. Not sure how to do that in pure CSS

File: src/components/mortgage-investment-compare/index.js [hunk 15]
```diff
@@ -23,17 +23,21 @@ import {
 const COMPOUND_FREQUENCY = 12
 
 class MortgageInvestmentCompare extends React.Component {
-  constructor () {
-    super()
+  constructor (props) {
+    super(props)
```
Comment (hdavidzhu) [inline src/components/mortgage-investment-compare/index.js]: What props are we giving this component?

File: src/components/mortgage-investment-compare/index.js [hunk 16]
```diff
@@ -152,12 +166,14 @@ class MortgageInvestmentCompare extends React.Component {
       }
     })
 
-    return {
+    const result = {
```
Comment (hdavidzhu) [inline src/components/mortgage-investment-compare/index.js]: sug: This could be directly returned

File: src/components/mortgage-investment-compare/input-card.js [hunk 17]
```diff
@@ -301,7 +330,7 @@ const InputCard = props => {
             the US during the 1960s was roughly $20,000. The average home price
             today is roughly{' '}
             <a href='https://www.census.gov/construction/nrs/pdf/uspricemon.pdf'>
-              $370,000
+              $400,000
```
Comment (hdavidzhu) [inline src/components/mortgage-investment-compare/input-card.js]: lol what happened

File: src/components/mortgage-investment-compare/index.js (hunk 15 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/index.js]: it's just a good practice and something I do every time I create a class. I've run into stupid bugs in the past for not doing so: https://overreacted.io/why-do-we-write-super-props/.

File: src/components/mortgage-investment-compare/index.js (hunk 16 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/index.js]: I like to name objects. I know it's obvious because the name of the function is `getResult` but for some reason when you first created this function I didn't think the function actually returned the whole result object in the state

File: src/components/mortgage-investment-compare/input-card.js (hunk 17 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/input-card.js]: I just updated it with the most recent data. Average price of a house has increased since I first wrote this. We'll make it dynamic eventually.

File: src/components/mortgage-investment-compare/amortization-table.js (hunk 8 above)
Comment (jgozal) [reply src/components/mortgage-investment-compare/amortization-table.js]: darnit david, just saw this comment. Making a new PR

File: src/components/mortgage-investment-compare/input-card.js (hunk 17 above)
Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/input-card.js]: wow. yikes.

File: src/components/mortgage-investment-compare/amortization-table.js (hunk 8 above)
Comment (hdavidzhu) [reply src/components/mortgage-investment-compare/amortization-table.js]: nah let's call it

Comment (jgozal) [reply src/components/mortgage-investment-compare/amortization-table.js]: NO I MUST REACH PERFECTION

Comment (jgozal) [reply src/components/mortgage-investment-compare/amortization-table.js]: just kidding, looked into it, it's a little too much work for the benefit
